dev/Pager/InteractionTests/PagerTests.cs

```diff
@@ -0,0 +1,960 @@
+            }
+        }
+
+        [TestMethod]
+        [TestProperty("TestSuite", "B")]
+        public void PreviousPageButtonChangingPageTest()
+        {
+            using (var setup = new TestSetupHelper("Pager Tests"))
+            {
+                elements = new PagerTestsPageElements();
+                VerifyPageChangedEventOutput(0);
+                InputHelper.LeftClick(elements.GetNextPageButton());
+
+                previousPage = 1;
+                InputHelper.LeftClick(elements.GetPreviousPageButton());
+                VerifyPageChangedEventOutput(0);
+
+                InputHelper.LeftClick(elements.GetLastPageButton());
+
+                previousPage = 4;
+                InputHelper.LeftClick(elements.GetPreviousPageButton());
+                VerifyPageChangedEventOutput(3);
+
+                InputHelper.LeftClick(elements.GetPreviousPageButton());
+                VerifyPageChangedEventOutput(2);
+
+                InputHelper.LeftClick(elements.GetPreviousPageButton());
+                VerifyPageChangedEventOutput(1);
+
+                InputHelper.LeftClick(elements.GetPreviousPageButton());
+                VerifyPageChangedEventOutput(0);
+
+            }
+        }
+
+        [TestMethod]
+        [TestProperty("TestSuite", "B")]
+        public void NextPageButtonChangingPageTest()
+        {
+            using (var setup = new TestSetupHelper("Pager Tests"))
+            {
+                elements = new PagerTestsPageElements();
+                VerifyPageChangedEventOutput(0);
+                InputHelper.LeftClick(elements.GetNextPageButton());
+                VerifyPageChangedEventOutput(1);
+
+                InputHelper.LeftClick(elements.GetNextPageButton());
+                VerifyPageChangedEventOutput(2);
+
+                InputHelper.LeftClick(elements.GetNextPageButton());
+                VerifyPageChangedEventOutput(3);
+
+                InputHelper.LeftClick(elements.GetNextPageButton());
+                VerifyPageChangedEventOutput(4);
+            }
+        }
+
+        [TestMethod]
+        [TestProperty("TestSuite", "A")]
+        public void LastPageButtonChangingPageTest()
+        {
+            using (var setup = new TestSetupHelper("Pager Tests"))
+            {
+                elements = new PagerTestsPageElements();
+                VerifyPageChangedEventOutput(0);
+                InputHelper.LeftClick(elements.GetLastPageButton());
+                VerifyPageChangedEventOutput(4);
+
+                InputHelper.LeftClick(elements.GetFirstPageButton());
+                InputHelper.LeftClick(elements.GetNextPageButton());
+
+                previousPage = 1;
+
+                InputHelper.LeftClick(elements.GetLastPageButton());
+                VerifyPageChangedEventOutput(4);
+            }
+        }
+
+        [TestMethod]
+        [TestProperty("TestSuite", "A")]
+        public void FirstPageButtonVisibilityOptionsTest()
```

Comment:
>FirstPageButtonVisibilityOptionsTest [](start = 20, length = 36)

There is a lot of duplicated code between this test and the Previous, next and last versions, can we factor out the test logic into a helper that  takes a button as a paramter and maybe an expected index shift?

Reply:
Agreed. I have refactored some logic to remove duplicated code!

Reply:
This is improved, however I think you could do better by using a parameterized base test, similar to this ScrollPresenter tests: [https://github.com/microsoft/microsoft-ui-xaml/blob/6265117611bfd5bb84a3c582c67071775fa9a1ca/dev/ScrollPresenter/InteractionTests/ScrollPresenterTestsWithInputHelper.cs#L273](https://github.com/microsoft/microsoft-ui-xaml/blob/6265117611bfd5bb84a3c582c67071775fa9a1ca/dev/ScrollPresenter/InteractionTests/ScrollPresenterTestsWithInputHelper.cs#L273)

---
In reply to: [463890133](https://github.com/microsoft/microsoft-ui-xaml/pull/2977#discussion_r463890133) [](ancestors = 463890133)

Reply:
I have added in a base test.